Rlabkey/R/labkey.rstudio.R

```diff
@@ -0,0 +1,194 @@
+##
+labkey.rstudio.saveReport <- function(folderPath, reportEntityId, reportFilename, useWarning=FALSE)
+{
+    ## check required parameters
+    if(missing(reportEntityId) || missing(reportFilename))
+        stop (paste("A value must be specified for each of reportEntityId and reportFilename."))
+
+    ## check working directory
+    if (!grepl(reportEntityId, getwd()))
+    {
+        return("Working directory is currently not set to report's directory. Skip saving source to LabKey.")
+    }
+
+    targetFilename <- labkey.rstudio.getSavedProp("reportFilename")
+    if (reportFilename != targetFilename)
+    {
+        return("Skip saving non LabKey report file.")
+    }
+
+    if(missing(folderPath))
+    {
+        folderPath <- labkey.rstudio.getSavedProp("folderPath")
+        if (is.null(folderPath) || folderPath == "NULL")
+        {
+            return("Unable to determine report folderPath. Skip saving source to LabKey.")
+        }
+    }
+
+    ## normalize the folder path
+    folderPath <- encodeFolderPath(folderPath)
+
+    baseUrl=labkey.getBaseUrl(NULL)
+
+    ## check valid report
+    url <- paste(baseUrl, "rstudio", folderPath, "ValidateRStudioReport.api", sep="")
+    params <- list(entityId=reportEntityId)
+    response <- labkey.post(url, toJSON(params, auto_unbox=TRUE))
+    lkResult <- (fromJSON(response))
+
+    if (lkResult$isValid == TRUE)
+    {
+        localLastModified <- labkey.rstudio.getSavedProp("lastModified")
+        doSave <- TRUE
+        if (localLastModified == lkResult$lastModified)
+        {
+            if (useWarning == TRUE)
+                doSave <- .rs.api.showQuestion("Save to LabKey?", "Do you want to update report content to LabKey Server?")
+        }
+        else
+        {
+            doSave <- .rs.api.showQuestion("Save to LabKey? (Potential conflicting edit)", "The report source was modified in LabKey Server and the content might have diverged from local copy. Do you want to save local changes to LabKey Server?")
+        }
+        if (!doSave)
+        {
+            return("Skipped saving updated source to LabKey server");
+        }
+        url <- paste(baseUrl, "rstudio", folderPath, "SaveRReportContent.api", sep="")
+
+        if (!file.exists(reportFilename))
+        {
+             stop (paste("File doesn't exist: ", reportFilename))
+        }
+        script <- readChar(reportFilename, file.info(reportFilename)$size)
+
+        params <- list(entityId=reportEntityId, runScript=script)
+        response <- labkey.post(url, toJSON(params, auto_unbox=TRUE))
+
+        result <- (fromJSON(response))
+
+        if (result$success == TRUE)
+        {
+            labkey.rstudio.updateProp("lastModified", result$lastModified)
+            return("Successfully updated report source to LabKey Server.")
+        }
+        else
+            warning(result$errorMsg)
+    }
+    else
+    {
+        warning("Failed to update report source to LabKey server. Report doesn't exist.")
```

Comment:
this is just a warning? seems like it should be an error / stop call like the others from above.